pkg/log/zap.go

```diff
@@ -16,26 +16,140 @@
 package log
 
 import (
+	"encoding/json"
 	"io"
+	"os"
+	"strconv"
+	"strings"
 
 	"github.com/go-logr/logr"
 	"github.com/go-logr/zapr"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
+const (
+	// Environment variable to control log format, if set to JSON, structured logging will be used. Default is "console", and to use zap.NewDevelopmentConfig.
+	LogFormat = "LOG_FORMAT"
+	// Log format environment variable value indicating JSON
+	LogFormatJSON = "json"
+	// Log format environment variable value indicating console / human-readable form
+	LogFormatConsole = "console"
+
+	// Environment variable to control logging level, should parse to zapcore.Level. Default is "info".
+	LogLevel = "LOG_LEVEL"
+
+	// Environment variable to control logging key used for log message in structured logging. Default is defined by zap.NewProductionEncoderConfig.
+	LogKeyMessage = "LOG_KEY_MESSAGE"
+
+	// Environment variable to control logging key used for log timestamp in structured logging. Default is defined by zap.NewProductionEncoderConfig.
+	LogKeyTimestamp = "LOG_KEY_TIMESTAMP"
+
+	// Environment variable to control the maximum field size in structured logging before truncating the field value.
+	LogMaxFieldSize = "LOG_MAX_FIELD_SIZE"
+
+	// Environment variable to control the logging timestamp format used in structured logging. Valid values are "rfc3339", "rfc3339nano" and "epoch".
+	LogTimestampFormat = "LOG_TIMESTAMP_FORMAT"
+)
+
 // NewDefaultZapLogger is the recommended way to create a new logger, you could call this function to initialize the root
 // logger of your application, and provide it to your components, by fx or manually.
 func NewDefaultZapLogger() (logr.Logger, error) {
-	// change the configuration in the future if needed.
-	zapLogger, err := zap.NewDevelopment()
+	logLevel := zap.InfoLevel
+	if envLevel := os.Getenv(LogLevel); envLevel != "" {
+		var err error
+		logLevel, err = zapcore.ParseLevel(envLevel)
+		if err != nil {
+			return logr.Discard(), err
+		}
+	}
+
+	logFormat := LogFormatConsole
+	if envFormat := os.Getenv(LogFormat); envFormat == LogFormatJSON {
+		logFormat = LogFormatJSON
+	}
+
+	var config zap.Config
+
+	if logFormat == LogFormatConsole {
+		config = zap.NewDevelopmentConfig()
+	} else {
+		encoderConfig := zap.NewProductionEncoderConfig()
+
+		if v := os.Getenv(LogKeyMessage); v != "" {
+			encoderConfig.MessageKey = v
+		}
+		if v := os.Getenv(LogKeyTimestamp); v != "" {
+			encoderConfig.TimeKey = v
+		}
+
+		if v := os.Getenv(LogTimestampFormat); v != "" {
+			var timeEncoder zapcore.TimeEncoder
+
+			if err := timeEncoder.UnmarshalText([]byte(v)); err != nil {
+				return logr.Discard(), err
```

Comment:
ditto, please fallback to a default timestamp format if we failed to parse that from ENV

Reply:
Interesting detail here is that `timeEncoder.UnmarshalText` in its current implementation never returns an error and silently defaults to `epoch`. I've added the error handler that just prints a warning, because it seems a bad idea to leave the error completely unhandled, but it won't be called with the current implementation in zap.

Could potentially drop the `else` clause that prints the error, given it's currently unreachable.